src/passes/userDefinedTypesConverter.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  ArrayTypeName,
+  ElementaryTypeName,
+  FunctionCall,
+  getNodeType,
+  Identifier,
+  Mapping,
+  MemberAccess,
+  UserDefinedType,
+  UserDefinedTypeName,
+  UserDefinedValueTypeDefinition,
+  VariableDeclaration,
+} from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+import { TranspileFailedError } from '../utils/errors';
+
+class UserDefinedValueTypeDefinitionEliminator extends ASTMapper {
+  visitUserDefinedValueTypeDefinition(node: UserDefinedValueTypeDefinition, _ast: AST): void {
+    node.vScope.removeChild(node);
+  }
+}
+
+export class UserDefinedTypesConverter extends ASTMapper {
+  visitVariableDeclaration(node: VariableDeclaration, ast: AST): void {
+    this.commonVisit(node, ast);
+    if (node.vType === undefined)
+      throw new TranspileFailedError(`Variable declaration type undefined for ${node.name}`);
+    if (node.typeString !== node.vType.typeString) node.typeString = node.vType.typeString;
+  }
+
+  visitArrayTypeName(node: ArrayTypeName, ast: AST): void {
+    this.commonVisit(node, ast);
+    if (node.vBaseType === undefined)
+      throw new TranspileFailedError(`Variable declaration type undefined for node id ${node.id}`);
+    if (node.typeString.slice(0, node.typeString.lastIndexOf('[')) !== node.vBaseType.typeString)
+      node.typeString =
+        node.vBaseType.typeString + node.typeString.substring(node.typeString.lastIndexOf('['));
+  }
+
+  visitMapping(node: Mapping, ast: AST): void {
+    this.commonVisit(node, ast);
+    node.typeString = `mapping(${node.vKeyType.typeString} => ${node.vValueType.typeString})`;
+  }
+
+  visitUserDefinedTypeName(node: UserDefinedTypeName, ast: AST): void {
+    const otherTypes: string[] = ['contract', 'enum', 'struct'];
+    if (otherTypes.some((type: string) => node.typeString.startsWith(type))) return;
```

Comment:
It's generally better to check whether the result of getNodeType is a UserDefinedType whose definition is a UserDefinedValueTypeDefinition or not rather than analysing the type strings like this

Reply:
Done